### PR TITLE
fix tuner.scale_batch_size not finding batch size attribute when using datamodule

### LINKED ISSUE #5967: "Wrong" code example in doc auto-scaling-of-batch-size section
## 🐛 Bug

In the doc [auto-scaling-of-batch-size section](https://pytorch-lightning.readthedocs.io/en/stable/training_tricks.html#auto-scaling-of-batch-size), a code example is 
```
# Use default in trainer construction
trainer = Trainer()
tuner = Tuner(trainer)

# Invoke method
new_batch_size = tuner.scale_batch_size(model, *extra_parameters_here)

# Override old batch size
model.hparams.batch_size = new_batch_size

# Fit as normal
trainer.fit(model)
```

However, this will not work as expected in the case where a LightningModule contains an attribute `self.datamodule`. Following the code will give  `MisconfigurationException: Field batch_size not found in both model and model.hparams`.

## To Reproduce
See my one-page code
```python
import torch
import torchvision
from torchvision import transforms
from torchvision import models
import utils
import pytorch_lightning as pl
from pytorch_lightning.callbacks import ModelCheckpoint

class ImageNetDataModule(pl.LightningDataModule):
    def __init__(self, imagenet_root, batch_size = 128, num_workers = 32):
        super().__init__()
        self.batch_size = batch_size
        self.imagenet_root = imagenet_root
        self.num_workers = num_workers
        
    def setup(self, stage):
        train_transform = transforms.Compose([transforms.Resize(256), 
                                        transforms.CenterCrop(224), 
                                        transforms.RandomGrayscale(),
                                        transforms.RandomHorizontalFlip(),
                                        transforms.ToTensor(), 
                                        transforms.Normalize(mean=[0.485, 0.456, 0.406],std=[0.229, 0.224, 0.225])])
        self.imagenet_train = torchvision.datasets.ImageNet(root = self.imagenet_root+"train/", split="train", transform = train_transform)
        val_transform = transforms.Compose([transforms.Resize(256), 
                                        transforms.CenterCrop(224), 
                                        transforms.ToTensor(), 
                                        transforms.Normalize(mean=[0.485, 0.456, 0.406],std=[0.229, 0.224, 0.225])])
        self.imagenet_val = torchvision.datasets.ImageNet(root= self.imagenet_root+"val/", split="val", transform = val_transform)
    
    def train_dataloader(self):
        return torch.utils.data.DataLoader(self.imagenet_train, 
                                           batch_size=self.batch_size,
                                           shuffle=True,
                                           num_workers = self.num_workers,
                                           persistent_workers=True
                                          )
    
    def val_dataloader(self):
        return torch.utils.data.DataLoader(self.imagenet_val, 
                                           batch_size=self.batch_size,
                                           shuffle=False,
                                           num_workers = self.num_workers,
                                           persistent_workers=True
                                          )
    
class NetWrapper(pl.LightningModule):
    def __init__(self, model,datamodule, criterion = torch.nn.CrossEntropyLoss()):
        super().__init__()
        self.model = model
        self.criterion = criterion
        self.lr = 1e-3
        self.datamodule = datamodule
    
    def forward(self, x):
        raw_prob = self.model(x) #(batch, 1000)
        return raw_prob
    
    def training_step(self, batch, batch_idx):
        x, y = batch
        preds = self(x) #(batch, word_emb_dim)
        loss = self.criterion(preds, y)
        self.log("cross_entropy_loss_training", loss)
        return loss
    
    def validation_step(self, batch, batch_idx):
        x, y = batch
        preds = self(x) #(batch, word_emb_dim)
        loss = self.criterion(preds, y)
        self.log("cross_entropy_loss_val", loss)
        
        return loss
    
    def validation_epoch_end(self, validation_step_outputs):
        all_outputs = torch.tensor(validation_step_outputs)
        std, mean = torch.std_mean(all_outputs)
        self.log("validation_mean", mean)
        self.log("validation_std", std)
        
    
    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=self.lr)
        return optimizer

imagenet_dm = ImageNetDataModule("../../datasets/ImageNet/", batch_size = 1024)
resnet50 = models.resnet50(pretrained=False)
resnet50_pl_module = NetWrapper(resnet50, imagenet_dm)
trainer = pl.Trainer(gpus=1, 
                     accelerator='dp', 
                     auto_scale_batch_size='binsearch')

tuner = pl.tuner.tuning.Tuner(trainer)
new_batch_size = tuner.scale_batch_size(resnet50_pl_module, mode="binsearch", init_val=128)
# the below line works fine
trainer.tune(resnet50_pl_module)
```

### Expected behavior

Tuner should find the attibute `batch_size` in `model.datamodule` in the method `Tuner.scale_batch_size()`.

### Environment

This issue should be independent of environments.

### Additional context

I took a look at the source code and found out that if we call `Trainer.tune()` directly, the invoke chain is 
`trainer.tune() -> tuner.tune()->tuner.scale_batch_size()->batch_size_scaling.scale_batch_size()->lightning_hasattr(model, attribute)->...` 
while the invoke chain of calling `Tuner.scale_batch_size()` is 
`tuner.scale_batch_size()->batch_size_scaling.scale_batch_size()->lightning_hasattr(model, attribute)->...`.
The problem is that `lightning_hasattr(model, attribute)` cannot find the attribute  `model.datamodule.batch_size` if we skip the registration steps in `trainer.tune()`.

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #5967

The tuner tries to detect if the model has an attribute for batch size, however before the trainer attaches the datamodule. 
Hence the batch size attribute can't be found. 

**The newly added test fails on master.**

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [ ] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by edenlightning** [comment]
@awaelchli any plans to get this in?

**Comment by awaelchli** [comment]
Yes I try to get it done by the end of this week.

File: tests/tuner/test_scale_batch_size.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        (BatchSizeModel(2), BatchSizeDataModule(2)),
+        (BatchSizeModel(2), BatchSizeDataModule(None)),
+        (BatchSizeModel(None), BatchSizeDataModule(2)),
+    ]
+)
+def test_scale_batch_size_method_with_model_or_datamodule(tmpdir, model, datamodule):
+    """ Test the tuner method `Tuner.scale_batch_size` with a datamodule. """
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        gpus=1,
```
Comment (carmocca) [inline tests/tuner/test_scale_batch_size.py:42]: why does the test need gpu?

**Review by carmocca — COMMENTED** [review]
Mildly related, why do we have this loop?

https://github.com/PyTorchLightning/pytorch-lightning/blob/34b733b35e02b13e01192a03ac4c3df89efa433a/pytorch_lightning/trainer/connectors/model_connector.py#L33

File: pytorch_lightning/tuner/tuning.py [hunk 2]
```diff
@@ -32,13 +32,22 @@ def on_trainer_init(self, auto_lr_find, auto_scale_batch_size):
...
+    ):
+        self.trainer.model_connector.copy_trainer_model_properties(model)
         # setup data, etc...
         self.trainer.train_loop.setup_fit(model, train_dataloader, val_dataloaders, datamodule)
-
         # hook
         self.trainer.data_connector.prepare_data(model)
 
+    def tune(self, model, train_dataloader, val_dataloaders, datamodule):
+        self.setup_trainer(model, train_dataloader, val_dataloaders, datamodule)
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/tuner/tuning.py:49]: also if someone has both `auto_scale_batch_size` and `auto_lr_find` enabled then also it will be called twice.

**Comment by awaelchli** [comment]
> Mildly related, why do we have this loop?

legacy stuff, pre- accelerator refactor
the loop goes over the lightning module and the potentially wrapped lightning module. 

this model connector class will be refactored-away.
In the end, the only thing we will ever want to set directly on the model is the trainer reference :) 
and all others can be derived through it

File: tests/tuner/test_scale_batch_size.py (hunk 1 above)
Comment (awaelchli) [reply tests/tuner/test_scale_batch_size.py]: oh you're right, it definitely doesn't.

File: pytorch_lightning/tuner/tuning.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/tuner/tuning.py:49]: yes but how to fix this? I don't see a good way with the current tuner design. 
We need to let the setup run every time because the user could do multiple tunings with different dataloaders etc.

[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/tuner/tuning.py]: I think it's fine since it's already called multiple times when `trainer.fit` is called with new_lr/new_batch_size while tuning. you can just remove the call from `.tune`

[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/tuner/tuning.py]: oh you already did :+1:

File: tests/tuner/test_scale_batch_size.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+import pytest
```
Comment (rohitgr7) [inline tests/tuner/test_scale_batch_size.py:1]: license

**Review by rohitgr7 — APPROVED** [review]
we have few tests related to tuner here: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/trainer/test_trainer_tricks.py

These should be added to tests/tuner too.

Else LGTM :v:

File: tests/tuner/test_scale_batch_size.py (hunk 3 above)
Comment (awaelchli) [reply tests/tuner/test_scale_batch_size.py:1]: thanks!

**Review by carmocca — APPROVED** [review]

**Review by s-rog — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Nice fix !

**Review by SkafteNicki — APPROVED** [review]
